src/main/java/org/sagebionetworks/bridge/services/UploadService.java

```diff
@@ -527,6 +568,97 @@ public void uploadComplete(String appId, UploadCompletionClient completedBy, Upl
         // Save uploadedOn date and uploadId to related adherence records.
         updateAdherenceWithUploadInfo(appId, upload);
     }
+
+    public void redriveUpload( byte[] fileBytes) throws IOException {
+        // Convert the byte[] array to a String
+        String fileContent = new String(fileBytes);
+
+        // Split the content into lines (each line is an entry - uploadId)
+        String[] lines = fileContent.split("\\r?\\n");
+
+        // Count the number of entries (lines)
+        int numberOfEntries = lines.length;
+
+        // Redrive upload.
+        if (numberOfEntries <= 10) {
+            redriveSmallAmountOfUploads(fileBytes);
+        } else {
+            redriveLargeAmountOfUploads(fileBytes);
+        }
+    }
+
+    private void redriveSmallAmountOfUploads(byte[] fileBytes) throws JsonProcessingException {
+        // Convert the byte[] array to a String
+        String fileContent = new String(fileBytes);
+
+        // Split the String into an array of lines
+        String[] linesArray = fileContent.split("\n");
+
+        // Convert the array of lines into a List
+        List<String> uploadIds = new ArrayList<>(Arrays.asList(linesArray));
+
+        for (String uploadId : uploadIds) {
+            Upload upload = getUpload(uploadId);
+            String appId = upload.getAppId();
+            if (appId == null) {
+                appId = healthCodeDao.getAppId(upload.getHealthCode());
+            }
+            uploadComplete(appId, UploadCompletionClient.REDRIVE, upload, true);
+            UploadValidationStatus validationStatus = pollUploadValidationStatusUntilComplete(uploadId);
+            if (validationStatus.getStatus() != UploadStatus.SUCCEEDED) {
+                logger.error("Redrive failed for uploadId=" + uploadId + ": " + COMMA_SPACE_JOINER.join(
+                        validationStatus.getMessageList()));
+            }
+        }
+    }
+
+    private void redriveLargeAmountOfUploads(byte[] fileBytes) throws IOException {
+        // set up s3 Key.
+        String currentYear = String.valueOf(DateUtils.getCurrentDateTime().getYear());
+        String currentMonth = String.valueOf(DateUtils.getCurrentDateTime().getMonthOfYear());
+        String s3Key = REDRIVE_UPLOAD_S3_KEY_PREFIX + currentYear + "-" + currentMonth;
```

Comment:
If two redrives happen at the same time for whatever reason, they will conflict with each other. It would be better to use DateUtils.getCurrentDateTime().getMillis().

If you need to test it, UploadServiceTest already uses Joda to mock the current time, so the time in the test is always TIMESTAMP.getMillis().